sw/device/lib/dif/dif_spi_device.c

```diff
@@ -0,0 +1,695 @@
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
+  }
+  switch (config->tx_order) {
+    case kDifSpiDeviceBitOrderMsbToLsb:
+      *device_config |= 0x0 << SPI_DEVICE_CFG_TX_ORDER;
+      break;
+    case kDifSpiDeviceBitOrderLsbToMsb:
+      *device_config |= 0x1 << SPI_DEVICE_CFG_TX_ORDER;
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
+  }
+  switch (config->rx_order) {
+    case kDifSpiDeviceBitOrderMsbToLsb:
+      *device_config |= 0x0 << SPI_DEVICE_CFG_RX_ORDER;
+      break;
+    case kDifSpiDeviceBitOrderLsbToMsb:
+      *device_config |= 0x1 << SPI_DEVICE_CFG_RX_ORDER;
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
+  }
+
+  *device_config = bitfield_set_field32(
+      *device_config, (bitfield_field32_t){
+                          .mask = SPI_DEVICE_CFG_TIMER_V_MASK,
+                          .index = SPI_DEVICE_CFG_TIMER_V_OFFSET,
+                          .value = config->rx_fifo_timeout,
+                      });
+
+  return kDifSpiDeviceResultOk;
+}
+
+dif_spi_device_result_t dif_spi_device_init(
+    mmio_region_t base_addr, const dif_spi_device_config_t *config,
+    dif_spi_device_t *spi) {
+  if (spi == NULL) {
+    return kDifSpiDeviceResultBadArg;
+  }
+  spi->base_addr = base_addr;
+
+  // NOTE: we do not write to any registers until performing all
+  // function argument checks, to avoid a halfway-configured SPI.
+
+  uint32_t device_config;
+  dif_spi_device_result_t control_error =
+      build_control_word(config, &device_config);
+  if (control_error != kDifSpiDeviceResultOk) {
+    return control_error;
+  }
+
+  uint16_t rx_fifo_start = 0x0;
+  uint16_t rx_fifo_end = config->rx_fifo_len - 1;
+  uint16_t tx_fifo_start = config->rx_fifo_len;
+  uint16_t tx_fifo_end = config->rx_fifo_len + config->tx_fifo_len - 1;
+  if (tx_fifo_end >= kDifSpiDeviceSramLen) {
+    // We've overflown the SRAM region...
+    return kDifSpiDeviceResultBadArg;
+  }
+
+  uint32_t rx_fifo_bounds = 0;
+  rx_fifo_bounds |= (rx_fifo_start & SPI_DEVICE_RXF_ADDR_BASE_MASK)
+                    << SPI_DEVICE_RXF_ADDR_BASE_OFFSET;
+  rx_fifo_bounds |= (rx_fifo_end & SPI_DEVICE_RXF_ADDR_LIMIT_MASK)
+                    << SPI_DEVICE_RXF_ADDR_LIMIT_OFFSET;
+
+  uint32_t tx_fifo_bounds = 0;
+  tx_fifo_bounds |= (tx_fifo_start & SPI_DEVICE_TXF_ADDR_BASE_MASK)
+                    << SPI_DEVICE_TXF_ADDR_BASE_OFFSET;
+  tx_fifo_bounds |= (tx_fifo_end & SPI_DEVICE_TXF_ADDR_LIMIT_MASK)
+                    << SPI_DEVICE_TXF_ADDR_LIMIT_OFFSET;
+
+  spi->rx_fifo_base = rx_fifo_start;
+  spi->rx_fifo_len = config->rx_fifo_len;
+  spi->tx_fifo_base = tx_fifo_start;
+  spi->tx_fifo_len = config->tx_fifo_len;
+
+  // Now that we know that all function arguments are valid, we can abort
+  // all current transactions and begin reconfiguring the device.
+  dif_spi_device_result_t err;
+  err = dif_spi_device_abort(spi);
+  if (err != kDifSpiDeviceResultOk) {
+    return err;
+  }
```

Comment:
Dead code. This error cannot occur since we now that `spi` cannot be `NULL` at this point.

Reply:
I'm not sure what the guidance is on not error-checking same-DIF functions called within other DIF functions...